### PR TITLE
Merge 2017-07-20 cda7381d880

### PR DESCRIPTION
**Pull request by rwhogg**
* [ ] lean-cli
* [ ] webpack

### CODE REVIEW

**Comment by maxim-belkin** [comment]
1. exiv2
```
checking for zlib.h... no
checking zlib in /usr... failed
configure: error: either specify a valid zlib installation with --with-zlib=DIR or disable zlib usage with --without-zlib
```

**Comment by maxim-belkin** [comment]
hbase
```
  [exec] /tmp/hbase--hadoop-lzo-20170721-51619-155z30w/hadoop-lzo-0.4.14/src/native/impl/lzo/LzoDecompressor.c: In function ‘Java_com_hadoop_compression_lzo_LzoDecompressor_initIDs’:
     [exec] /tmp/hbase--hadoop-lzo-20170721-51619-155z30w/hadoop-lzo-0.4.14/src/native/impl/lzo/LzoDecompressor.c:90:37: error: expected expression before ‘,’ token
     [exec]   liblzo2 = dlopen(HADOOP_LZO_LIBRARY, RTLD_LAZY | RTLD_GLOBAL);
     [exec]                                      ^
     [exec] /tmp/hbase--hadoop-lzo-20170721-51619-155z30w/hadoop-lzo-0.4.14/src/native/impl/lzo/LzoCompressor.c: In function ‘Java_com_hadoop_compression_lzo_LzoCompressor_initIDs’:
     [exec] /tmp/hbase--hadoop-lzo-20170721-51619-155z30w/hadoop-lzo-0.4.14/src/native/impl/lzo/LzoCompressor.c:125:37: error: expected expression before ‘,’ token
     [exec]   liblzo2 = dlopen(HADOOP_LZO_LIBRARY, RTLD_LAZY | RTLD_GLOBAL);
     [exec]                                      ^
     [exec] /tmp/hbase--hadoop-lzo-20170721-51619-155z30w/hadoop-lzo-0.4.14/src/native/impl/lzo/LzoCompressor.c: In function ‘Java_com_hadoop_compression_lzo_LzoCompressor_compressBytesDirect’:
     [exec] /tmp/hbase--hadoop-lzo-20170721-51619-155z30w/hadoop-lzo-0.4.14/src/native/impl/lzo/LzoCompressor.c:274:3: warning: implicit declaration of function ‘strstr’ [-Wimplicit-function-declaration]
     [exec]    } else if (strstr(lzo_compressor_function, "lzo1x_999")
     [exec]    ^
     [exec] /tmp/hbase--hadoop-lzo-20170721-51619-155z30w/hadoop-lzo-0.4.14/src/native/impl/lzo/LzoCompressor.c:274:14: warning: incompatible implicit declaration of built-in function ‘strstr’ [enabled by default]
     [exec]    } else if (strstr(lzo_compressor_function, "lzo1x_999")
     [exec]               ^
     [exec] make: *** [impl/lzo/LzoDecompressor.lo] Error 1
     [exec] make: *** Waiting for unfinished jobs....
     [exec] make: *** [impl/lzo/LzoCompressor.lo] Error 1

BUILD FAILED
/tmp/hbase--hadoop-lzo-20170721-51619-155z30w/hadoop-lzo-0.4.14/build.xml:256: exec returned: 2
```

**Comment by maxim-belkin** [comment]
qt
```
==> ./configure -verbose -prefix /home/linuxbrew/.linuxbrew/Cellar/qt/5.9.1 -release -opensource -confirm-license -system-zlib -qt-libpng -qt-libjpeg -qt-freetype -qt-pcre -nomake tests -pkg-config -dbus-runtime -qt-xcb -R/home/linuxbrew/.linuxbrew/Cellar/qt/5.9.1/lib -nomake examples
...
Last 150 lines from /home/ubuntu/.cache/Homebrew/Logs/qt/01.configure:
  libproxy ............................... no
  OpenSSL ................................ yes
    Qt directly linked to OpenSSL ........ no
  SCTP ................................... no
  Use system proxies ..................... yes
Qt Gui:
  Accessibility .......................... yes
  FreeType ............................... yes
    Using system FreeType ................ no
  HarfBuzz ............................... yes
    Using system HarfBuzz ................ no
  Fontconfig ............................. no
  Image formats:
    GIF .................................. yes
    ICO .................................. yes
    JPEG ................................. yes
      Using system libjpeg ............... no
    PNG .................................. yes
      Using system libpng ................ no
  EGL .................................... no
  OpenVG ................................. no
  OpenGL:
    Desktop OpenGL ....................... no
    OpenGL ES 2.0 ........................ no
    OpenGL ES 3.0 ........................ no
    OpenGL ES 3.1 ........................ no
  Session Management ..................... yes
Features used by QPA backends:
  evdev .................................. yes
  libinput ............................... no
  INTEGRITY HID .......................... no
  mtdev .................................. no
  tslib .................................. no
  xkbcommon-evdev ........................ yes
QPA backends:
  DirectFB ............................... no
  EGLFS .................................. no
  LinuxFB ................................ yes
  VNC .................................... yes
  Mir client ............................. no
  X11:
    Using system-provided XCB libraries .. no
    EGL on X11 ........................... no
    Xinput2 .............................. yes
    XCB XKB .............................. yes
    XLib ................................. yes
    XCB render ........................... yes
    XCB GLX .............................. yes
    XCB Xlib ............................. yes
    Using system-provided xkbcommon ...... yes
Qt Widgets:
  GTK+ ................................... no
  Styles ................................. Fusion Windows
Qt PrintSupport:
  CUPS ................................... no
Qt Sql:
  DB2 (IBM) .............................. no
  InterBase .............................. no
  MySql .................................. no
  OCI (Oracle) ........................... no
  ODBC ................................... no
  PostgreSQL ............................. no
  SQLite2 ................................ no
  SQLite ................................. yes
    Using system provided SQLite ......... no
  TDS (Sybase) ........................... no
Qt SerialBus:
  Socket CAN ............................. yes
  Socket CAN FD .......................... yes
QtXmlPatterns:
  XML schema support ..................... yes
Qt QML:
  QML interpreter ........................ yes
  QML network support .................... yes
Qt Quick:
  Direct3D 12 ............................ no
  AnimatedImage item ..................... yes
  Canvas item ............................ yes
  Support for Qt Quick Designer .......... yes
  Flipable item .......................... yes
  GridView item .......................... yes
  ListView item .......................... yes
  Path support ........................... yes
  PathView item .......................... yes
  Positioner items ....................... yes
  ShaderEffect item ...................... yes
  Sprite item ............................ yes
Qt Gamepad:
  SDL2 ................................... no
Qt 3D:
  Assimp ................................. yes
  System Assimp .......................... no
  Output Qt3D Job traces ................. no
  Output Qt3D GL traces .................. no
Qt 3D GeometryLoaders:
  Autodesk FBX ........................... no
Qt Wayland Client ........................ no
Qt Wayland Compositor .................... no
Qt Bluetooth:
  BlueZ .................................. no
  BlueZ Low Energy ....................... no
  Linux Crypto API ....................... no
Qt Sensors:
  sensorfw ............................... no
Qt Multimedia:
  ALSA ................................... no
  GStreamer 1.0 .......................... no
  GStreamer 0.10 ......................... no
  Video for Linux ........................ yes
  OpenAL ................................. no
  PulseAudio ............................. no
  Resource Policy (libresourceqt5) ....... no
  Windows Audio Services ................. no
  DirectShow ............................. no
  Windows Media Foundation ............... no
Qt Quick Controls 2:
  Styles ................................. Default Material Universal
Qt Quick Templates 2:
  Hover support .......................... yes
  Multi-touch support .................... yes
Qt Location:
  Gypsy GPS Daemon ....................... no
  WinRT Geolocation API .................. no
Qt WebEngine:
  Embedded build ......................... no
  Pepper Plugins ......................... yes
  Printing and PDF ....................... yes
  Proprietary Codecs ..................... no
  Spellchecker ........................... yes
  WebRTC ................................. yes
  Using system ninja ..................... no
  ALSA ................................... no
  PulseAudio ............................. no
  System libraries:
    ICU .................................. no
    libwebp and libwebpdemux ............. no
    Opus ................................. no
    ffmpeg ............................... no

Note: Also available for Linux: linux-clang linux-icc

Note: Dropped compiler flags '-pthread' when detecting library 'glib'.

Note: Disabling Linux Accessibility Bridge: D-Bus is missing.

Note: No wayland-egl support detected. Cross-toolkit compatibility disabled.

ERROR: The OpenGL functionality tests failed!
You might need to modify the include and library search paths by editing QMAKE_INCDIR_OPENGL[_ES2],
QMAKE_LIBDIR_OPENGL[_ES2] and QMAKE_LIBS_OPENGL[_ES2] in the mkspec for your platform.
```

**Comment by maxim-belkin** [comment]
qscintilla2
```
==> brew audit qscintilla2 --online
Error: 4 problems in 1 formula
qscintilla2:
  * Dependency sip should not use option with-python3
  * Dependency pyqt should not use option with-python
  * Dependency pyqt should not use option with-python
  * Dependency sip should not use option with-python3
==> FAILED
```

**Comment by maxim-belkin** [comment]
gcc
```
==> Downloading https://linuxbrew.bintray.com/bottles/gcc-5.3.0.x86_64_linux.bottle.tar.gz
/usr/bin/curl --remote-time --location --user-agent Linuxbrew/1.2.4-2-g59a338502 (Linux; x86_64 Ubuntu 14.04.4 LTS) curl/7.35.0 --fail --silent https://linuxbrew.bintray.com/bottles/gcc-5.3.0.x86_64_linux.bottle.tar.gz -C 0 -o /home/ubuntu/.cache/Homebrew/gcc-5.3.0.x86_64_linux.bottle.tar.gz.incomplete
Error: Failed to download resource "gcc"
Download failed: https://linuxbrew.bintray.com/bottles/gcc-5.3.0.x86_64_linux.bottle.tar.gz
```

**Comment by rwhogg** [comment]
> checking for zlib.h... no

Thanks for adding the `zlib` dependency Maxim 😸 

> * stable: sha256 changed without the version also changing; please create an issue upstream to rule out malicious circumstances and to find out why the file changed.

That was done upstream by https://github.com/Exiv2/exiv2/issues/19, which offered an acceptable explanation. It's not malicious.

>   * Dependency sip should not use option with-python3
  * Dependency pyqt should not use option with-python
  * Dependency pyqt should not use option with-python
  * Dependency sip should not use option with-python3

As I personally don't have the time to fix these at the moment, I'm okay with leaving those as is for now. 

> Error: Failed to download resource "gcc"

Yikes! I hope that was just a temporary network fluke. It seemed fine when I tried to download it today.

**Comment by rwhogg** [comment]
`exiv2` on Travis: https://travis-ci.org/Linuxbrew/homebrew-core/builds/256052934#L4598

```
configure: error: either specify a valid expat installation with --with-expat=DIR or disable XMP support with --disable-xmp
```

Locally `brew linkage exiv2` says `expat` and `zlib` are the only deps missing.

**Comment by maxim-belkin** [comment]
so we need to add `depends_on "expat"`?

**Comment by rwhogg** [comment]
Yeah, was about to do just that.

**Comment by maxim-belkin** [comment]
wait a second

**Comment by maxim-belkin** [comment]
I'm adding a fix for hbase

**Comment by rwhogg** [comment]
Sorry, didn't read that in time and already pushed

**Comment by maxim-belkin** [comment]
```
=> brew install --build-bottle --verbose aescrypt-packetizer
==> Downloading https://www.aescrypt.com/download/v3/linux/aescrypt-3.13.tgz
Already downloaded: /home/ubuntu/.cache/Homebrew/aescrypt-packetizer-3.13.tgz
==> Verifying aescrypt-packetizer-3.13.tgz checksum
tar xzf /home/ubuntu/.cache/Homebrew/aescrypt-packetizer-3.13.tgz
==> make
.
Last 150 lines from /home/ubuntu/.cache/Homebrew/Logs/aescrypt-packetizer/01.make:
2017-07-21 15:03:04 +0000

make

gcc -Wall -Wextra -pedantic -std=c99 -D_FILE_OFFSET_BITS=64 -c aescrypt.c
gcc -Wall -Wextra -pedantic -std=c99 -D_FILE_OFFSET_BITS=64 -c aes.c
gcc -Wall -Wextra -pedantic -std=c99 -D_FILE_OFFSET_BITS=64 -c sha256.c
gcc -Wall -Wextra -pedantic -std=c99 -D_FILE_OFFSET_BITS=64 -c password.c
gcc -Wall -Wextra -pedantic -std=c99 -D_FILE_OFFSET_BITS=64 -c keyfile.c
gcc -Wall -Wextra -pedantic -std=c99 -D_FILE_OFFSET_BITS=64   -c -o aescrypt_keygen.o aescrypt_keygen.c
gcc -Wall -Wextra -pedantic -std=c99 -D_FILE_OFFSET_BITS=64 -liconv -o aescrypt_keygen aescrypt_keygen.o keyfile.o password.o
/usr/bin/ld: cannot find -liconv
```

File: Formula/hbase.rb [hunk 1]
```diff
@@ -19,6 +19,7 @@ class Hbase < Formula
   # the lzo jar has a native extension
   # building native extensions requires a version of java that matches the architecture
   # there is no 32 bit version of java for macOS since Java 1.7, and 1.7+ is required for hbase
+  depends_on "gcc" unless OS.mac?
```
Comment (maxim-belkin) [inline Formula/hbase.rb]: this is probably should be `:build` :/

**Comment by maxim-belkin** [comment]
```
 $ brew linkage hbase
System libraries:
  /lib/x86_64-linux-gnu/libc.so.6
  /lib/x86_64-linux-gnu/libdl.so.2
Missing libraries:
  libjvm.so
```

**Comment by maxim-belkin** [comment]
`/home/linuxbrew/.linuxbrew/Cellar/jdk/1.8.0-131/jre/lib/amd64/server/libjvm.so`

File: Formula/hbase.rb (hunk 1 above)
Comment (rwhogg) [inline Formula/hbase.rb]: I was thinking something like `fails_with :gcc => "4.8"` might be appropriate.

**Comment by rwhogg** [comment]
Hmm, maybe we need to change `brew linkage` so it checks for .so files recursively to find cases like libjvm.so

File: Formula/hbase.rb (hunk 1 above)
Comment (maxim-belkin) [inline Formula/hbase.rb]: with `fails_with :gcc => "4.8"` but without `depends_on "gcc"` I get:
```
$ brew install hbase
Error: hbase cannot be built with any available compilers.
Install Clang or brew install gcc
```

**Comment by maxim-belkin** [comment]
well, I think `linkage` works fine:
```
$ ldd /home/linuxbrew/.linuxbrew/Cellar/hbase/1.2.6/libexec/lib/native/Linux-amd64-64/libgplcompression.so.0.0.0
	linux-vdso.so.1 =>  (0x00007ffcb6588000)
	libjvm.so => not found
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fc1a5f55000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fc1a5b8d000)
	/lib64/ld-linux-x86-64.so.2 (0x0000559101eb5000)
```

**Comment by rwhogg** [comment]
> /usr/bin/ld: cannot find -liconv

`inreplace "Makefile", "#LIBS=-liconv", "LIBS=-liconv"` probably needs to be guarded by `OS.mac?`. There's a similar case in the HEAD build, but I'm not personally as worried about that ATM.

**Comment by maxim-belkin** [comment]
Interestingly enough, I see the following in the logs for hbase:
```
[exec] libtool: link: gcc-5 -shared  impl/lzo/.libs/LzoCompressor.o impl/lzo/.libs/LzoDecompressor.o   -L/home/linuxbrew/.linuxbrew/opt/jdk/jre/lib/amd64/server -L/home/linuxbrew/.linuxbrew/opt/jdk/lib/amd64/server -ljvm -ldl  -m64 -m64   -Wl,-soname -Wl,libgplcompression.so.0 -o .libs/libgplcompression.so.0.0.0
```

**Comment by maxim-belkin** [comment]
I can reproduce the failure of `gcc-4.8` with `gcc-5.3` by adding the following line to `hbase`:
```
ENV["LDFLAGS"] = "-Wl,--as-needed"
```
after `ENV["CPPFLAGS"] = ...`

**Comment by rwhogg** [comment]
Thanks a bunch for all the fixes provided @maxim-belkin!

**Pull request merged**